libbuteosyncfw/common/Logger.cpp

```diff
@@ -21,215 +22,38 @@
-#include <QFileInfo>
-#include <QDir>
 
-#include <syslog.h>
+#define TRACE_FUNCTION_TIME 0
 
 using namespace Buteo;
 
-const int Logger::DEFAULT_INDENT_SIZE = 4;
-
-Logger *Logger::sInstance = nullptr;
-
-Logger *Logger::instance()
-{
-    if (nullptr == sInstance) {
-        createInstance();
-    }
-    return sInstance;
-}
-
-int Logger::defaultLogLevel()
-{
-    QByteArray levelByteArray = qgetenv("MSYNCD_LOGGING_LEVEL");
-    QString levelStr = QString::fromLocal8Bit(levelByteArray.constData());
-    bool ok = false;
-    int level = levelStr.toInt(&ok);
-    if (!ok) {
-        level = 4;  // warning level
-    }
-    return level;
-}
-
-void Logger::createInstance(const QString &aLogFileName, bool aUseStdOut,
-                            int aIndentSize)
-{
-    deleteInstance();
-    sInstance = new Logger(aLogFileName, aUseStdOut, aIndentSize);
-}
-
-void Logger::deleteInstance()
-{
-    delete sInstance;
-    sInstance = nullptr;
-}
-
-Logger::Logger(const QString &aLogFileName, bool aUseStdOut, int aIndentSize)
-    :   iEnabledLevels(NUM_LEVELS),
-        iIndentLevel(0),
-        iIndentSize(aIndentSize),
-        iFileStream(0),
-        iStdOutStream(0),
-        iStdErrStream(0),
-        iEnabled(false),
-        iLogLevel(0)
+LogTimer::LogTimer(const QString &categoryName, const QString &func)
+    : m_categoryName(categoryName.toUtf8())
+    , m_func(func)
+    , m_category(m_categoryName.constData())
 {
-    iLogLevel = defaultLogLevel();
-
-    if (aUseStdOut) {
-        // Create a stream for writing log messages to standard output.
-        iStdOutStream = new QTextStream(stdout);
-    }
-
-    iStdErrStream = new QTextStream(stderr);
-
-    if (!aLogFileName.isEmpty()) {
-        // Create stream for writing log messages to a file.
-        iFile.setFileName(aLogFileName);
-        QFileInfo fileInfo(iFile);
-        QDir dir;
-        dir.mkpath(fileInfo.absolutePath());
-        if (iFile.open(QIODevice::WriteOnly | QIODevice::Append)) {
-            iFileStream = new QTextStream(&iFile);
-        }
-    }
+    qCDebug(m_category) << m_func << ":Entry";
+    m_timer.start();
 }
 
-Logger::~Logger()
+LogTimer::~LogTimer()
 {
-    delete iFileStream;
-    iFileStream = 0;
-
-    delete iStdOutStream;
-    iStdOutStream = 0;
-
-    delete iStdErrStream;
-    iStdErrStream = 0;
+    qCDebug(m_category) << m_func << ":Exit, execution time:" << m_timer.elapsed() << "ms";
 }
 
-
-bool  Logger::setLogLevel(int aLevel)
+void LogTimer::log(const QLoggingCategory &category, const QString &func)
 {
-    bool retVal = false;
-    if ((aLevel > 0)  && (aLevel <= NUM_LEVELS)) {
-        iLogLevel = aLevel;
-        disable(iEnabledLevels);
-        QBitArray iLevels(NUM_LEVELS, false);
-        for (int i = aLevel; i > 0; i--) {
-            iLevels.setBit(NUM_LEVELS - i);
-        }
-        enable(iLevels);
-        retVal = true;
+    if (category.isDebugEnabled()) {
+        QScopedPointer<Buteo::LogTimer> timerDebugVariable;
```

Comment:
won't this immediately go out of scope, rather than going out of scope at the end of the function?
Maybe I'm wrong, but I think instead of having this static ::log() function, this should just be implemented in the macro definition:

```
#define FUNCTION_CALL_TRACE(loggingCategory) \
    QScopedPointer<Buteo::LogTimer> timerDebugVariable( \
            loggingCategory.isDebugEnabled() \
            ? new Buteo::LogTimer(QString::fromUtf8(loggingCategory.categoryName(), QString(Q_FUNC_INFO)) \
            : nullptr)
```

Reply:
Good point, updated it to use the recommended code.